quil-rs/src/expression.rs

```diff
@@ -253,6 +254,592 @@ fn is_small(x: f64) -> bool {
+                    Ok(expr.add(node))
+                }
+                Sexp::List(list) if list.is_empty() => Err(SimplificationError::EmptySexp),
+                Sexp::List(list) => match &list[0] {
+                    Sexp::Empty => Err(SimplificationError::EmptySexp), // should be unreachable
+                    list @ Sexp::List(..) => {
+                        Err(SimplificationError::UnexpectedListSexp(list.clone()))
+                    }
+                    Sexp::String(op) => {
+                        let arg_ids: Vec<Id> = list[1..]
+                            .iter()
+                            .map(|s| parse_sexp_into(s, expr))
+                            .collect::<Result<_, _>>()?;
+                        let node = Expr::from_op(op, arg_ids)
+                            .map_err(|e| SimplificationError::UnknownOp(format!("{e:#?}")))?;
+                        Ok(expr.add(node))
+                    }
+                },
+            }
+        }
+        let mut expr = RecExpr::default();
+        parse_sexp_into(sexp, &mut expr)?;
+        Ok(expr)
+    }
+
+    /// Our analysis will perform arithmetic simplification (largely, constant folding) on our
+    /// language.
+    #[derive(Default)]
+    struct Arithmetic;
+    type EGraph = egg::EGraph<Expr, Arithmetic>;
+
+    /// Our analysis will perform constant folding on our language.
+    impl egg::Analysis<Expr> for Arithmetic {
+        /// Constant values
+        type Data = Option<Complex>;
+
+        /// Pull the (possible) [`Self::Data`] from the given expression.
+        fn make(egraph: &EGraph, enode: &Expr) -> Self::Data {
+            let x = |id: &Id| egraph[*id].data.as_ref();
+            match enode {
+                Expr::Pi => Some(Complex::PI),
+                Expr::Number(c) => Some(*c),
+                Expr::Cis(id) => Some(x(id)?.cis()),
+                Expr::Cos(id) => Some(x(id)?.cos()),
+                Expr::Exp(id) => Some(x(id)?.exp()),
+                Expr::Sin(id) => Some(x(id)?.sin()),
+                Expr::Sqrt(id) => Some(x(id)?.sqrt()),
+                Expr::Pos(id) => Some(*x(id)?),
+                Expr::Neg(id) => Some(-*x(id)?),
+                Expr::Pow([base, power]) => Some(x(base)?.pow(*x(power)?)),
+                Expr::Mul([left, right]) => Some(*x(left)? * *x(right)?),
+                Expr::Div([left, right]) => Some(*x(left)? / *x(right)?),
+                Expr::Add([left, right]) => Some(*x(left)? + *x(right)?),
+                Expr::Sub([left, right]) => Some(*x(left)? - *x(right)?),
+                Expr::Address(_) | Expr::Symbol(_) => None,
+            }
+        }
+
+        /// Merge two pieces of data with the same value.
+        fn merge(&mut self, to: &mut Self::Data, from: Self::Data) -> egg::DidMerge {
+            egg::merge_option(to, from, |a, b| {
+                assert!(a.close(b), "Merged non-equal constants: {a} & {b}");
```

Comment:
Can we avoid this panic?

Reply:
Sure! The only times it's been triggered in loads of randomized testing is when dealing with `NaN` and `inf`; those worries are currently avoided with tighter bounds on what floats are allowed in the property-based testing.